pkg/services/ngalert/migration/service.go

```diff
@@ -0,0 +1,200 @@
+package migration
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	pb "github.com/prometheus/alertmanager/silence/silencepb"
+
+	"github.com/grafana/grafana/pkg/infra/db"
+	"github.com/grafana/grafana/pkg/infra/kvstore"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const KVNamespace = "ngalert.migration"
+const migratedKey = "migrated"
+
+type MigrationService struct {
+	store db.DB
+	cfg   *setting.Cfg
+	log   log.Logger
+	kv    *kvstore.NamespacedKVStore
+}
+
+func ProvideService(
+	cfg *setting.Cfg,
+	sqlStore db.DB,
+	kv kvstore.KVStore,
+) (*MigrationService, error) {
+	return &MigrationService{
+		log:   log.New("ngalert.migration"),
+		cfg:   cfg,
+		store: sqlStore,
+		kv:    kvstore.WithNamespace(kv, 0, KVNamespace),
+	}, nil
+}
+
+// Run starts the migration.
+func (ms *MigrationService) Run(ctx context.Context) error {
+	ms.log.Info("Starting")
+	migrated, err := ms.GetMigrated(ctx)
+	if err != nil {
+		return fmt.Errorf("getting migration status: %w", err)
+	}
+	if migrated == ms.cfg.UnifiedAlerting.IsEnabled() {
+		// Nothing to do.
+		ms.log.Info("Nothing to do")
+		return nil
+	}
+
+	if migrated {
+		// If legacy alerting is also disabled, there is nothing to do
+		if setting.AlertingEnabled != nil && !*setting.AlertingEnabled {
+			return nil
+		}
+
+		// Safeguard to prevent data loss when migrating from UA to LA
+		if !ms.cfg.ForceMigration {
+			panic("Grafana has already been migrated to Unified Alerting.\nAny alert rules created while using Unified Alerting will be deleted by rolling back.\n\nSet force_migration=true in your grafana.ini and restart Grafana to roll back and delete Unified Alerting configuration data.")
```

Comment:
I think it can be a regular error. The service is running as part of the error Group anyway.

Reply:
Ah, true. Nice catch!